tools/Google.Cloud.Tools.Analyzers/SymbolExtensions.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using Microsoft.CodeAnalysis;
+
+namespace Google.Cloud.Tools.Analyzers
+{
+    internal static class SymbolExtensions
+    {
+        internal static ITypeSymbol GetVariableType(this ISymbol symbol)
+        {
+            switch (symbol)
+            {
+                case ILocalSymbol local: return local.Type;
+                case IParameterSymbol parameter: return parameter.Type;
+                case IFieldSymbol field: return field.Type;
+                default: throw new ArgumentException("The specified symbol is not a variable.", nameof(symbol));
+            }
+        }
+
+        internal static bool IsExternallyVisible(this ISymbol symbol)
+        {
+            switch (symbol.DeclaredAccessibility)
+            {
+                case Accessibility.Public:
+                case Accessibility.Protected:
+                case Accessibility.ProtectedOrInternal:
+                    return symbol.ContainingType?.IsExternallyVisible() ?? true;
+                default:
+                    return false;
+            }
+        }
+
+        internal static bool IsPrimitive(this ITypeSymbol type)
```

Comment:
Are you concerned about performance or just the code style? If the former, I don't think you can get much better than what's here. The compiler seems to create range checks on blocks of contiguous values:

```il
.method assembly hidebysig static bool  IsPrimitive(class [Microsoft.CodeAnalysis]Microsoft.CodeAnalysis.ITypeSymbol 'type') cil managed
{
  .custom instance void [mscorlib]System.Runtime.CompilerServices.ExtensionAttribute::.ctor() = ( 01 00 00 00 ) 
  // Code size       25 (0x19)
  .maxstack  2
  .locals init (valuetype [Microsoft.CodeAnalysis]Microsoft.CodeAnalysis.SpecialType V_0)
  IL_0000:  ldarg.0
  IL_0001:  callvirt   instance valuetype [Microsoft.CodeAnalysis]Microsoft.CodeAnalysis.SpecialType [Microsoft.CodeAnalysis]Microsoft.CodeAnalysis.ITypeSymbol::get_SpecialType()
  IL_0006:  stloc.0
  IL_0007:  ldloc.0
  IL_0008:  ldc.i4.7
  IL_0009:  sub
  IL_000a:  ldc.i4.s   9
  IL_000c:  ble.un.s   IL_0015
  IL_000e:  ldloc.0
  IL_000f:  ldc.i4.s   18
  IL_0011:  sub
  IL_0012:  ldc.i4.1
  IL_0013:  bgt.un.s   IL_0017
  IL_0015:  ldc.i4.1
  IL_0016:  ret
  IL_0017:  ldc.i4.0
  IL_0018:  ret
} // end of method SymbolExtensions::IsPrimitive
```

However, if it's just about style, I can change it.